util/reggen/gen_tock.py

```diff
@@ -0,0 +1,384 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""
+Generate Rust constants from validated register JSON tree
+"""
+
+import io
+import logging as log
+import sys
+import textwrap
+import warnings
+from datetime import datetime
+from typing import Any, Dict, List, Optional, Set, TextIO
+
+from .ip_block import IpBlock
+from .multi_register import MultiRegister
+from .params import LocalParam
+from .register import Register
+from .window import Window
+
+REG_VISIBILITY = 'pub(crate)'
+FIELD_VISIBILITY = 'pub(crate)'
+
+
+def genout(outfile: TextIO, msg: str, *args: Any, **kwargs: Any) -> None:
+    outfile.write(msg.format(*args, **kwargs))
+
+
+def indent(s: str, amount: int = 4) -> str:
+    result = []
+    indent = 0
+    for line in s.splitlines():
+        indented = []
+        comment = False
+        for i, ch in enumerate(line):
+            if i > 0 and ch == '/' and line[i - 1] == '/':
+                comment = True
+            if not comment and ch in ")]}":
+                indent -= 1
+            if i == 0:
+                indented.append(' ' * indent * amount)
+            if not comment and ch in "([{":
+                indent += 1
+            indented.append(ch)
+        result.append(''.join(indented))
+        if line == '':
+            result.append('')
+    return '\n'.join(result)
+
+
+def to_snake_case(s: str) -> str:
+    val = []
+    for i, ch in enumerate(s):
+        if i > 0 and ch.isupper():
+            val.append('_')
+        val.append(ch)
+    return ''.join(val)
+
+
+def to_studly_caps(s: str) -> str:
+    words = [w.capitalize() for w in s.split('_')]
+    return ''.join(words)
+
+
+def to_upper_snake_case(s: str) -> str:
```

Comment:
Add function comments to `to_studly_caps` and `to_upper_snake_case` demonstrating the case they produce. From reading the code, I think they produce `To_Studly_Caps` and `TO_UPPER_SNAKE_CASE`, but I'm not sure.

Edit after reading code that uses `to_studly_caps`: Oh, `to_studly_caps` produces UpperCamelCase. Is studly caps a common term? I've always seen it called UpperCamelCase.

Reply:
Done.

I'd heard it referred to as "StudlyCaps" once.  Considering the association between Java and camelCase, I prefer to avoid the term camelCase (and lets not talk about PascalCase).

Unfortunately, there are differing opinions on what "StudlyCaps" actually means:
- https://stackoverflow.com/questions/32731717/what-is-the-difference-between-studlycaps-and-camelcase
- https://en.wikipedia.org/wiki/Alternating_caps